### PR TITLE
feat!: add muon optimizer support

### PR DESCRIPTION
**Pull request by kinggongzilla**
This PR adds support for Muon optimizer, as well as a recipe and comparison of AB-UPT with Lion vs Muon.

Non-2D weights are inherently not supported by muon, hence they need to be updated by a secondary optimizer.

Note that qkv weight matrix had to be split up into separate q, k and v matrices, in order for muon to work. Hence some tests had to be adjusted. **This also breaks checkpoint loading for previously saved models!**

Todos:

- [x] Add tests
- [x] Add muon recipe with results
- [x] Add readme with results

### CODE REVIEW

File: src/noether/core/optimizer/optimizer_wrapper.py [hunk 1]
```diff
@@ -158,6 +158,11 @@ def _apply_parameter_modifiers(
             # excluding norm and bias params is very common for all models -> support with simple flag
             # bias has ndim == 1, so it needs to be checked before
             # the bias of norm layers is considered a bias, not a norm parameter
+            if param.ndim >= 2:
+                properties["use_muon"] = True
+            else:
+                properties["use_muon"] = False
+
```
Comment (Ndles) [inline src/noether/core/optimizer/optimizer_wrapper.py]: can be a oneliner: `properties["use_muon"] = param.ndim >= 2`

File: src/noether/core/optimizer/optimizer_wrapper.py [hunk 2]
```diff
@@ -201,6 +206,10 @@ def _add_names_to_param_groups(self, merged_groups: list[dict[str, Any]]) -> lis
             for key, value in param_group.items():
                 if key == "params":
                     continue
+                if key == "kind":
+                    continue
+                if key == "use_muon":
+                    continue
```
Comment (Ndles) [inline src/noether/core/optimizer/optimizer_wrapper.py]: can be a single check: 
```python
if key in ["params", "kind", "use_muon"]:
    continue
```

File: src/noether/core/optimizer/optimizer_wrapper.py [hunk 3]
```diff
@@ -275,8 +284,10 @@ def step(self, grad_scaler: GradScaler | None = None) -> None:
             grad_scaler.unscale_(self.torch_optim)
         # clip gradients
         if self.config.clip_grad_value is not None:
+            assert self.all_parameters is not None
```
Comment (Ndles) [inline src/noether/core/optimizer/optimizer_wrapper.py]: we prefer to raise an error in the code, asserts are for tests

File: src/noether/modeling/modules/attention/dot_product.py [hunk 4]
```diff
@@ -59,13 +61,24 @@ def forward(
...
+            "bs seqlen (num_heads head_dim) -> bs num_heads seqlen head_dim",
+            num_heads=self.num_heads,
+            head_dim=self.head_dim,
+        )
+        v = einops.rearrange(
+            self.v(x),
+            "bs seqlen (num_heads head_dim) -> bs num_heads seqlen head_dim",
+            num_heads=self.num_heads,
+            head_dim=self.head_dim,
+        )
```
Comment (HennerM) [inline src/noether/modeling/modules/attention/dot_product.py]: I think it's nice to keep the qkv projection as one matmul, 
can we do something like this:

```suggestion
        qkv_weight = torch.cat([self.q.weight, self.k.weight, self.v.weight], dim=0)
        qkv_bias = torch.cat([self.q.bias, self.k.bias, self.v.bias], dim=0) if self.q.bias is not None else None
        qkv = F.linear(x, qkv_weight, qkv_bias)
        
        q, k, v = einops.rearrange(
            qkv,
            "bs seqlen (three num_heads head_dim) -> three bs num_heads seqlen head_dim",
            three=3,
            num_heads=self.num_heads,
            head_dim=self.head_dim,
        ).unbind(0)
```

File: src/noether/modeling/modules/attention/dot_product.py [hunk 5]
```diff
@@ -38,7 +38,9 @@ def __init__(
         self.dropout = config.dropout
         self.proj_dropout = nn.Dropout(config.dropout)
 
-        self.qkv = nn.Linear(config.hidden_dim, config.hidden_dim * 3, bias=config.bias)
+        self.q = nn.Linear(config.hidden_dim, config.hidden_dim, bias=config.bias)
+        self.k = nn.Linear(config.hidden_dim, config.hidden_dim, bias=config.bias)
+        self.v = nn.Linear(config.hidden_dim, config.hidden_dim, bias=config.bias)
```
Comment (HennerM) [inline src/noether/modeling/modules/attention/dot_product.py:43]: This breaks checkpoint compatibility (which is quite annoying), maybe this is something we want to mention in the PR/changelog

File: src/noether/modeling/modules/attention/perceiver.py [hunk 6]
```diff
@@ -90,14 +91,20 @@ def forward(
             # Project K/V from input
             if kv is None:
                 raise ValueError("Either kv or kv_cache must be provided.")
-            kv_proj = self.kv(kv)
-            k, v = einops.rearrange(
-                kv_proj,
-                "bs seqlen_kv (two num_heads head_dim) -> two bs num_heads seqlen_kv head_dim",
-                two=2,
+            k = self.k(kv)
```
Comment (HennerM) [inline src/noether/modeling/modules/attention/perceiver.py]: same here with the concat matmul

File: src/noether/core/optimizer/optimizer_wrapper.py [hunk 7]
```diff
@@ -158,6 +158,8 @@ def _apply_parameter_modifiers(
             # excluding norm and bias params is very common for all models -> support with simple flag
             # bias has ndim == 1, so it needs to be checked before
             # the bias of norm layers is considered a bias, not a norm parameter
+            properties["use_muon"] = param.ndim >= 2
```
Comment (HennerM) [inline src/noether/core/optimizer/optimizer_wrapper.py]: Do we have to add this in here? Can we not do the same check inside `muon_composite` itself?

File: src/noether/core/schemas/optimizers.py [hunk 8]
```diff
@@ -37,6 +37,23 @@ def check_scale_or_value_exclusive(self) -> Self:
...
+    """The learning rate for the optimizer. Falls back to the primary lr if not set."""
+    weight_decay: float | None = Field(None, ge=0.0)
+    """The weight decay. Falls back to the primary weight_decay if not set."""
+    momentum: float | None = Field(None, ge=0.0, le=1.0)
+    """Momentum factor for optimizers like SGD."""
+    betas: tuple[float, float] | None = None
+    """Beta coefficients for Adam-style optimizers."""
+
+
 class OptimizerConfig(BaseModel):
```
Comment (HennerM) [inline src/noether/core/schemas/optimizers.py:57]: Can we have a `MuonOptimizerConfig` instead of adding the attributes in here?

File: src/noether/core/schemas/optimizers.py [hunk 9]
```diff
@@ -61,6 +78,14 @@ class OptimizerConfig(BaseModel):
     weight_decay_schedule: AnyScheduleConfig | None = Field(None, discriminator="kind")
     schedule_config: AnyScheduleConfig | None = Field(None, discriminator="kind")
 
+    momentum: float | None = Field(None, ge=0.0, le=1.0)
+    """Momentum factor for optimizers like SGD and Muon."""
+    betas: tuple[float, float] | None = None
+    """Beta coefficients for Adam-style optimizers."""
+    secondary: MuonSecondaryOptimizerConfig | None = None
```
Comment (HennerM) [inline src/noether/core/schemas/optimizers.py]: These all sound like they should be separate schemas:
* AdamOptimizerConfig
* SGDOptimizerConfig
* MuonOptimizerConfig

having momentum and beta is confusing because Adam's beta are the first and second momentum. If the configs are split apart this makes more sense though.

**Review by HennerM — COMMENTED** [review]
I think we should split up the optimizer configs a bit, having all options in the same schema takes some of the advantage of schemas.

Otherwise the logic looks fine. The composition is cool and the Muon+Lion experiments idea is nice too!

File: src/noether/modeling/modules/attention/dot_product.py (hunk 5 above)
Comment (kinggongzilla) [reply src/noether/modeling/modules/attention/dot_product.py:43]: true! i actually didnt think about that. I put it in the PR description for now

File: recipes/aero_cfd/configs/experiment/drivaerml/ab_upt_muon_lion.yaml [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    num_volume_anchor_points: 16384
+
+inference_pipeline:
+  num_geometry_points: 65536
+  num_geometry_supernodes: 16384
+  num_surface_anchor_points: 1000000000
+  num_volume_anchor_points: 1000000000
+
+trainer:
+  precision: bfloat16
```
Comment (HennerM) [inline recipes/aero_cfd/configs/experiment/drivaerml/ab_upt_muon_lion.yaml]: why is this not float16?

Comment (kinggongzilla) [reply recipes/aero_cfd/configs/experiment/drivaerml/ab_upt_muon_lion.yaml]: good question. in some initial shapenet experiments with Muon training showed some large loss spikes with float16 and more stable/lower loss with bfloat16. So I stuck with this for following experiments

Comment (HennerM) [reply recipes/aero_cfd/configs/experiment/drivaerml/ab_upt_muon_lion.yaml]: did you train your shapenet baselines with bfloat16 as well? I think @MauritsBleeker noticed in the past that actually float16 converges to a lower loss, but I don't think we ever did a proper analysis on precision and numerical stability

Comment (kinggongzilla) [reply recipes/aero_cfd/configs/experiment/drivaerml/ab_upt_muon_lion.yaml]: ~yes, the shapenet baselines were with bfloat16, after noticing these training instabilities with flaot16 with muon~
@HennerM sorry this is wrong, I misread your question. the baselines were trained with the default config with float16

File: src/noether/core/schemas/optimizers.py (hunk 9 above)
Comment (kinggongzilla) [reply src/noether/core/schemas/optimizers.py]: I now split up the schemas, such that OptimizerConfig schema contains shared arguments between optimizers and other optimizer schemas hold specific arguments and inherit from it.

File: src/noether/core/schemas/optimizers.py [hunk 11]
```diff
@@ -73,3 +96,32 @@ class OptimizerConfig(BaseModel):
 
     def return_optim_wrapper_args(self) -> dict:
         return self.model_dump(include=self._optim_wrapper_kwargs)
+
+
+class AdamOptimizerConfig(OptimizerConfig):
```
Comment (HennerM) [inline src/noether/core/schemas/optimizers.py:101]: nice this supports both Adam and Lion :)

**Review by HennerM — APPROVED** [review]
Code wise this looks fine to me now, good work!
Up to you if you want to wait for more experiments to come in or merge this change already

**Comment by kinggongzilla** [comment]
@claude review

**Comment by kinggongzilla** [comment]
@claude review

File: src/noether/core/optimizer/muon_composite.py [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+        # Create internal Muon optimizer, forwarding only explicitly set kwargs
+        muon_kwargs: dict = dict(lr=lr, momentum=momentum, weight_decay=weight_decay)
+        if nesterov is not None:
+            muon_kwargs["nesterov"] = nesterov
+        if ns_steps is not None:
+            muon_kwargs["ns_steps"] = ns_steps
+        if adjust_lr_fn is not None:
+            muon_kwargs["adjust_lr_fn"] = adjust_lr_fn
+        self._muon = torch.optim.Muon(muon_groups, **muon_kwargs) if muon_groups else None
+        from noether.core.factory.utils import class_constructor_from_class_path
```
Comment (Ndles) [inline src/noether/core/optimizer/muon_composite.py:76]: I am not a huge fan of lazy imports, is there a reason for it? Also, initializing a class within a class here feels a bit weird, would it make sense to init MuonComposite with the already initialized instance instead?

File: src/noether/core/optimizer/muon_composite.py
Comment (Ndles) [inline src/noether/core/optimizer/muon_composite.py:1]: I think here `def state_dict()` and `def load_state_dict()` can be optimized a bit, but probably not a big deal.

**Review by Ndles — APPROVED** [review]
A few comments, but nothing critical.
